sky/packages/sky/lib/src/rendering/custom_layout.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'box.dart';
+import 'object.dart';
+
+class _MultiChildParentData extends ContainerBoxParentDataMixin<RenderBox> { }
+
+abstract class MultiChildLayoutDelegate {
+  Map<Object, RenderBox> _idToChildMap = new Map<Object, RenderBox>();
+  RenderBox _idToChild(Object childId) { return  _idToChildMap[childId]; }
+
+  Size layoutChild(Object childId, BoxConstraints constraints) {
+    final RenderBox child = _idToChild(childId);
+    child.layout(constraints, parentUsesSize: true);
+    return child.size;
+  }
+
+  void positionChild(Object childId, Point position) {
+    final RenderBox child = _idToChild(childId);
+    final _MultiChildParentData childParentData = child.parentData;
+    childParentData.position = position;
+  }
+
+  void _callPerformLayout(Size size, BoxConstraints constraints, List<Object> childIds, RenderBox firstChild) {
+    RenderBox child = firstChild;
+    int index = 0;
+    while (child != null) {
+      _idToChildMap[childIds[index++]] = child;
+      final _MultiChildParentData childParentData = child.parentData;
+      child = childParentData.nextSibling;
+    }
+
+    performLayout(size, constraints);
+
+    _idToChildMap.clear();
+  }
+
+  void performLayout(Size size, BoxConstraints constraints);
+}
+
+class RenderCustomMultiChildLayoutBox extends RenderBox
+  with ContainerRenderObjectMixin<RenderBox, _MultiChildParentData>,
+       RenderBoxContainerDefaultsMixin<RenderBox, _MultiChildParentData> {
+  RenderCustomMultiChildLayoutBox({
+    List<RenderBox> children,
+    List<Object> childKeys,
+    MultiChildLayoutDelegate delegate
+  }) : _delegate = delegate, _childKeys = childKeys {
+    assert(delegate != null);
+    addAll(children);
+  }
+
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _MultiChildParentData)
+      child.parentData = new _MultiChildParentData();
+  }
+
+  MultiChildLayoutDelegate get delegate => _delegate;
+  MultiChildLayoutDelegate _delegate;
+  void set delegate (MultiChildLayoutDelegate newDelegate) {
+    assert(newDelegate != null);
+    if (_delegate == newDelegate)
+      return;
+    _delegate = newDelegate;
+    markNeedsLayout();
+  }
+
+  List<Object> get childKeys => _childKeys;
+  List<Object> _childKeys;
+  void set childKeys (List<Object> newChildrenKeys) {
+    // TBD: verify that children.length == childKeys.length
+    if (_childKeys == newChildrenKeys)
+      return;
+    _childKeys = newChildrenKeys;
+    markNeedsLayout();
+  }
+
+  Size _getSize(BoxConstraints constraints) {
+    return constraints.constrain(constraints.biggest);
+  }
```

Comment:
IMHO, we should put this function on the delegate, like we did for the one child case.  That way the client can determine the size.